rust/connlib/tunnel/src/tests.rs

```diff
@@ -2595,6 +2615,37 @@ fn assert_dns_packets_properties(state: &TunnelTest, ref_state: &ReferenceState)
     }
 }
 
+fn pick_dns_resource<'a>(
+    name: &str,
+    resources: impl IntoIterator<Item = &'a ResourceDescriptionDns>,
+) -> Option<ResourceId> {
+    resources
+        .into_iter()
+        .filter(|r| is_subdomain(name, &r.address))
+        .sorted_by(|r1, r2| Ord::cmp(&r2.address.len(), &r1.address.len()))
+        .map(|r| r.id)
+        .next()
+}
+
+fn is_subdomain(name: &str, record: &str) -> bool {
+    if name == record {
+        return true;
+    }
+    let Some((first, end)) = record.split_once('.') else {
+        return false;
+    };
+    match first {
+        "*" => name.ends_with(end) && name.strip_suffix(end).is_some_and(|n| n.ends_with('.')),
+        "?" => {
+            name.ends_with(end)
+                && name
+                    .strip_suffix(end)
+                    .is_some_and(|n| n.ends_with('.') && n.matches('.').count() == 1)
+        }
+        _ => false,
+    }
+}
```

Comment:
There is already a utility called `matches_domain` that should help you here!

Reply:
`matches_domain` is a bit more limited, since it matches something like `?.foo` to `bar.baz.foo`

Reply:
maybe we should replace `matches_domain` with this?

Reply:
> maybe we should replace `matches_domain` with this?

Yeah. Can we implement it in a simpler way? Or with less duplication?

Reply:
maybe it's better to use `is_subdomain` that's implemented in the dns module? it's quite simple and it has some separated unit-tests so it's not that circular

Reply:
Hmm, I guess it is okay? I didn't really want to depend on implementation details like this in the test too but if it is separately tested should be fine.